packages/modules/notification/src/services/notification-module-service.ts

```diff
@@ -108,57 +109,114 @@ export default class NotificationModuleService
     const idempotencyKeys = data
       .map((entry) => entry.idempotency_key)
       .filter(Boolean)
-    const alreadySentNotifications = await this.notificationService_.list(
-      {
-        idempotency_key: idempotencyKeys,
-      },
-      { take: null },
-      sharedContext
-    )
 
-    const existsMap = new Map(
-      alreadySentNotifications.map((n) => [n.idempotency_key as string, true])
-    )
+    let { notificationsToProcess, createdNotifications } =
+      await this.baseRepository_.transaction(async (txManager) => {
+        const context = {
+          ...sharedContext,
+          transactionManager: txManager,
+        }
 
-    const notificationsToProcess = data.filter(
-      (entry) => !entry.idempotency_key || !existsMap.has(entry.idempotency_key)
-    )
+        const alreadySentNotifications = await this.notificationService_.list(
+          {
+            idempotency_key: idempotencyKeys,
+          },
+          { take: null },
+          context
+        )
 
-    const notificationsToCreate = await promiseAll(
-      notificationsToProcess.map(async (entry) => {
-        const provider =
-          await this.notificationProviderService_.getProviderForChannel(
-            entry.channel
-          )
+        const existsMap = new Map(
+          alreadySentNotifications.map((n) => [n.idempotency_key as string, n])
+        )
 
-        if (!provider) {
-          throw new MedusaError(
-            MedusaError.Types.NOT_FOUND,
-            `Could not find a notification provider for channel: ${entry.channel}`
-          )
-        }
+        const notificationsToProcess = data.filter(
+          (entry) =>
+            !entry.idempotency_key ||
+            !existsMap.has(entry.idempotency_key) ||
+            (existsMap.has(entry.idempotency_key) &&
+              [NotificationStatus.PENDING, NotificationStatus.FAILURE].includes(
+                existsMap.get(entry.idempotency_key)!.status
+              ))
+        )
 
-        if (!provider.is_enabled) {
-          throw new MedusaError(
-            MedusaError.Types.NOT_FOUND,
-            `Notification provider ${provider.id} is not enabled. To enable it, configure it as a provider in the notification module options.`
+        const channels = notificationsToProcess.map((not) => not.channel)
+        const providers =
+          await this.notificationProviderService_.getProviderForChannels(
+            channels
           )
-        }
 
-        const res = await this.notificationProviderService_.send(
-          provider,
-          entry
+        // Create the notifications to be sent to prevent concurrent actions listing the same notifications
+        const normalizedNotificationsToProcess = notificationsToProcess.map(
```

Comment:
suggestion:
Instead of `alreadyExists` flag, should we just set the `idempotencyKey` as part of the returned object, and then below in the `create` call just do `filter(e => entry.idempotency_key && existsMap.has(entry.idempotency_key)`? I think it's a bit easier to follow, but it might just be me.

Feel free to go with whatever you think is better.